service-monitor/lib/terminate-cdn-service.js

```diff
@@ -0,0 +1,33 @@
+/**
+ * @param {{
+ *   DB: D1Database
+ *   TRANSACTION_QUEUE: import('cloudflare:workers').Queue<{
+ *     dataSetId: number
+ *   }>
+ * }} env
+ */
+export async function terminateCDNServiceForSanctionedWallets(env) {
+  const { results: dataSets } = await env.DB.prepare(
+    `
+      SELECT DISTINCT ds.id
+      FROM data_sets ds
+        LEFT JOIN wallet_details wd ON ds.payer_address = wd.address
+      WHERE ds.with_cdn = 1 AND wd.is_sanctioned = 1;
+  `,
+  ).run()
```

Comment:
It will take some time until the TXs terminating the datasets are accepted by the network, Goldsky picks up the termination events and invokes our webhook.

During that time, our scheduled worker will be called, possibly more than once. Those calls will pick up the same set of datasets to terminate, and create more transactions with duplicate "terminate service" calls for the same data set ID.

IMO, we need to keep track of which datasets have a pending termination TX, and ignore them in this query.

Reply:
Yes that's true. While the CF Workflows can solve such problems using their unique IDs it's not possible to solve this problem using queues without persisting state in either durable object or database. 

Another (temporary) solution would be to have this worker triggered at lower frequency (i.e. every 10 minutes).

Reply:
> IMO, we need to keep track of which datasets have a pending termination TX, and ignore them in this query.

Added `terminate_service_tx_hash ` column to data sets in https://github.com/filcdn/worker/pull/243/commits/2b4c23c635a61bc9b8294ec4fa079f4895129834 and https://github.com/filcdn/worker/pull/243/commits/7b4681e98cbf37723bf2f21f98a5ff9a4f283582